### PR TITLE
Add header file docs for openssl/core_numbers.h and openssl/core_names.h

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by mspncp** [comment]
As a side note related to this pr: looking at the content of `<openssl/core_numbers.h>` I wonder whether the header name `<openssl/core_functions.h>` wouldn't be more appropriate? It's all about defining provider _functions_ and the _number_, (i.e. the function ID) is just a part of their definition.

https://github.com/openssl/openssl/blob/e847085914476636d75ee1874b78e1c0e983da6e/include/openssl/core_numbers.h#L90-L128

Also, there are parts in the header where the IDs are intertwined with the function declarations, and other parts where the IDs precede the declarations blockwise.

https://github.com/openssl/openssl/blob/e847085914476636d75ee1874b78e1c0e983da6e/include/openssl/core_numbers.h#L130-L150

I find the latter variant more readable. Whichever variant you choose, it should be consistent.

**Comment by Tymbur** [comment]
I don’t know why I received this it looks like it belongs to someone else
and it happened because of the github iOS app it looks like a security
issue to me there are enumerated blobs and shit in here .  This is me
saying report this error .  I don’t know if this were someone else besides
me if it would be a big deal or not but it looks like a personal email I’m
not sure.

**Comment by mspncp** [comment]
@Tymbur I have no idea why you received the mail. It was a regular post of mine on pull request #11963. To  prevent further email notifications, you can follow the link to the pull request and unsubscribe from the thread.

**Comment by Tymbur** [comment]
Sure thing just wanted to make sure it wasn’t something serious

**Comment by levitte** [comment]
> As a side note related to this pr: looking at the content of `<openssl/core_numbers.h>` I wonder whether the header name `<openssl/core_functions.h>` wouldn't be more appropriate?

Not sure that's the best replacement names either, the name "core_functions" would make me expect to see a number of declarations for functions offered by the core of libcrypto...  regarding the name "core_numbers", I would argue that these *are* dispatch numbers offered by the core, along with the function types that the provider need to associate with those numbers.

**Comment by levitte** [comment]
> Also, there are parts in the header where the IDs are intertwined with the function declarations, and other parts where the IDs precede the declarations blockwise.
> ...
> I find the latter variant more readable. Whichever variant you choose, it should be consistent.

I would say that it depends.  The functions offered by the core (!!!) is a looong list, that's a lot of jumping back and forth to see what function number macro corresponds to what function signature.
Although, one might argue that some operations have a long list of numbers/function signatures as well, where we separate the number block from the function signature block, so we're not quite consistent there.

**Comment by levitte** [comment]
Now that I've commented on stuff that isn't about this PR itself at all, anyone want to review?

**Comment by slontis** [comment]
There are some make doc-nits errors in travis.

File: doc/man7/openssl-core_names.h.pod [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+openssl/core_names.h - OpenSSL provider parameter names
+
+=head1 SYNOPSIS
+
+ #include <openssl/core_names.h>
+
+=head1 DESCRIPTION
+
+The F<< <openssl/core_names.h> >> header defines a multitude of macros
+for L<OSSL_PARAM(3)> names, algorithm names and other known names used
```
Comment (slontis) [inline doc/man7/openssl-core_names.h.pod:14]: It also contains some param values..

File: doc/man7/openssl-core_names.h.pod [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+with OpenSSL's providers, made available for practical purposes only.
+
+Existing names are further described in the manuals for OpenSSL's
+providers (see L</SEE ALSO>) and the manuals for each algorithm they
+provide (listed in those provider manuals).
+
+=head1 SEE ALSO
+
+L<OSSL_PROVIDER-default(7)>, L<OSSL_PROVIDER-FIPS(7)>
```
Comment (slontis) [inline doc/man7/openssl-core_names.h.pod:23]: legacy and NULL also?

File: doc/man7/openssl-core_names.h.pod [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+=head1 HISTORY
+
+The macros described here were added in OpenSSL 3.0.
+
+=head1 CAVEATS
+
+=begin comment
+
+CAVEATS is a section name borrowed from OpenBSD, where mdoc(7)
```
Comment (slontis) [inline doc/man7/openssl-core_names.h.pod:33]: Is this comment really needed?

File: doc/man7/openssl-core_names.h.pod [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+=end comment
+
+I<This header file does not constitute a general registry of names>.
+Providers that implement new algorithms are to be responsible for
+their own parameter names.
+
+However, authors of provider that implement their own variants of
+algorithms that OpenSSL providers support will want to pay attention
+to the names provided in this header to work in a compatible manner.
+
```
Comment (slontis) [inline doc/man7/openssl-core_names.h.pod:49]: missing the stuff normally at the ned 

=cut

File: doc/man7/openssl-core_numbers.h.pod [hunk 5]
```diff
@@ -0,0 +1,56 @@
+=pod
+
+=head1 NAME
+
+openssl/core_numbers.h
+- OpenSSL provider interface numbers and function types
```
Comment (slontis) [inline doc/man7/openssl-core_numbers.h.pod:6]: for numbers maybe?

File: doc/man7/openssl-core_names.h.pod (hunk 3 above)
Comment (levitte) [reply doc/man7/openssl-core_names.h.pod:33]: I know some nit-picky people that are likely to ask.

File: doc/man7/openssl-core_names.h.pod (hunk 4 above)
Comment (levitte) [reply doc/man7/openssl-core_names.h.pod:49]: Ah, thanks!

File: doc/man7/openssl-core_numbers.h.pod (hunk 5 above)
Comment (levitte) [reply doc/man7/openssl-core_numbers.h.pod:6]: Er, no?  But perhaps "dispatch numbers" is better than "interface numbers"

File: doc/man7/openssl-core_names.h.pod (hunk 1 above)
Comment (levitte) [reply doc/man7/openssl-core_names.h.pod:14]: Aren't they "other known names"?

File: doc/man7/openssl-core_names.h.pod (hunk 3 above)
Comment (slontis) [reply doc/man7/openssl-core_names.h.pod:33]: Yeah I understand why you did it :)
Maybe move to the git comments instead?

File: doc/man7/openssl-core_names.h.pod (hunk 1 above)
Comment (slontis) [reply doc/man7/openssl-core_names.h.pod:14]: ok

File: doc/man7/openssl-core_numbers.h.pod (hunk 5 above)
Comment (slontis) [reply doc/man7/openssl-core_numbers.h.pod:6]: identifiers and function wrappers?

File: doc/man7/openssl-core_names.h.pod (hunk 2 above)
Comment (levitte) [reply doc/man7/openssl-core_names.h.pod:23]: The null provider doesn't provide much, so I don't see why it should be mentioned here.  I've added the legacy provider to the list, though.

File: doc/man7/openssl-core_numbers.h.pod (hunk 5 above)
Comment (levitte) [reply doc/man7/openssl-core_numbers.h.pod:6]: ... especially since I use "dispatch number" everywhere else.  Done

Comment (levitte) [reply doc/man7/openssl-core_numbers.h.pod]: > identifiers and function wrappers?

Those are new terms that we have never used before in this context...  and "wrappers"?  that's doesn't connect with what they are, at least to me.

Comment (slontis) [reply doc/man7/openssl-core_numbers.h.pod]: Ok.. I cant think of anything better.

File: doc/man7/openssl-core_names.h.pod (hunk 3 above)
Comment (levitte) [reply doc/man7/openssl-core_names.h.pod]: Oh, I forgot about this.  I did as you suggested

**Comment by levitte** [comment]
Care to re-approve with the squash commit I just added?

**Comment by richsalz** [comment]
I dislike adding a new CAVEATS section.  OpenSSL has always used NOTES, which seems quite appropriate here.

Don't blaze new trails.  Be boring.  Be  consistent.

**Comment by levitte** [comment]
You often want to collapse NOTES into DESCRIPTION too... :wink:

I think this particular one does deserve a bit more "voice" than a mere note...

**Comment by richsalz** [comment]
Yes, I often collapse NOTES into DESCRIPTION because it doesn't deserve a separate call-out, and folks get lazy by just adding NOTES after the main manpage is written.

Look at other notes sections.  This is no different.

I'm sure I'll lose this point, sigh, but you're still wrong.

**Comment by levitte** [comment]
Merged.

329b2a2cde480fbfe1cf43bf8ecf1c75a16cf6bf DOCS: add openssl-core_numbers.h(7)
f438f53a4e57462216be271c1c965550b6ff9941 DOCS: add openssl-core_names.h(7)

**Pull request closed without merging**